### PR TITLE
Worker profile cleanup

### PR DESCRIPTION
**Pull request by liliankasem**
Just tidying up as I work on the active profile bug fix, figured better to check this in separately

### CODE REVIEW

File: src/WebJobs.Script/Workers/Profiles/Conditions/HostPropertyCondition.cs [hunk 1]
```diff
@@ -68,27 +67,27 @@ public bool Evaluate()
                 return false;
             }
 
-            _logger.LogDebug($"Evaluating HostPropertyCondition with name: {Name}, value: {value} and expression {Expression}");
+            _logger.LogDebug("Evaluating HostPropertyCondition with name: {name}, value: {value} and expression {expression}", Name, value, Expression);
```
Comment (kshyju) [inline src/WebJobs.Script/Workers/Profiles/Conditions/HostPropertyCondition.cs:70]: Should HostProperty enum be public ?

File: src/WebJobs.Script/Workers/Rpc/Configuration/RpcWorkerConfigFactory.cs [hunk 2]
```diff
@@ -286,31 +298,33 @@ internal bool ShouldAddWorkerConfig(string workerDescriptionLanguage)
...
                 // After specialization only create worker provider for the language set by FUNCTIONS_WORKER_RUNTIME env variable
-                _logger.LogInformation($"{RpcWorkerConstants.FunctionWorkerRuntimeSettingName} set to {_workerRuntime}. Skipping WorkerConfig for language:{workerDescriptionLanguage}");
+                _logger.LogInformation("{RpcWorkerConstants.FunctionWorkerRuntimeSettingName} set to {workerRuntime}. Skipping WorkerConfig for language: {}", RpcWorkerConstants.FunctionWorkerRuntimeSettingName, _workerRuntime, workerDescriptionLanguage);
                 return false;
             }
+
             return true;
         }
 
         internal void ReadLanguageWorkerFile(string workerPath)
```
Comment (kshyju) [inline src/WebJobs.Script/Workers/Rpc/Configuration/RpcWorkerConfigFactory.cs]: Should this method be private?

**Review by kshyju — APPROVED** [review]

File: src/WebJobs.Script/Workers/Profiles/Conditions/HostPropertyCondition.cs (hunk 1 above)
Comment (liliankasem) [reply src/WebJobs.Script/Workers/Profiles/Conditions/HostPropertyCondition.cs:70]: I don't see why it is, I can make it private!

File: src/WebJobs.Script/Workers/Rpc/Configuration/RpcWorkerConfigFactory.cs [hunk 3]
```diff
@@ -165,21 +171,25 @@ internal void AddProvider(string workerDir)
...
                         };
+
                         _workerDescriptionDictionary[workerDescription.Language] = rpcWorkerConfig;
                         ReadLanguageWorkerFile(arguments.WorkerPath);
-                        _logger.LogDebug($"Added WorkerConfig for language: {workerDescription.Language}");
+
+                        _logger.LogDebug("Added WorkerConfig for language: {language}", workerDescription.Language);
                     }
                 }
                 catch (Exception ex)
```
Comment (jviau) [inline src/WebJobs.Script/Workers/Rpc/Configuration/RpcWorkerConfigFactory.cs]: Can we add a `when (!ex.IsFatal())` here?

File: src/WebJobs.Script/Workers/Rpc/Configuration/RpcWorkerConfigFactory.cs [hunk 4]
```diff
@@ -286,31 +298,33 @@ internal bool ShouldAddWorkerConfig(string workerDescriptionLanguage)
...
-                _logger.LogDebug($"EnvironmentVariable {RpcWorkerConstants.FunctionWorkerRuntimeSettingName}: {_workerRuntime}");
+                _logger.LogDebug("EnvironmentVariable {functionWorkerRuntimeSettingName}: {workerRuntime}", RpcWorkerConstants.FunctionWorkerRuntimeSettingName, _workerRuntime);
                 if (_workerRuntime.Equals(workerDescriptionLanguage, StringComparison.OrdinalIgnoreCase))
                 {
                     return true;
                 }
+
                 // After specialization only create worker provider for the language set by FUNCTIONS_WORKER_RUNTIME env variable
-                _logger.LogInformation($"{RpcWorkerConstants.FunctionWorkerRuntimeSettingName} set to {_workerRuntime}. Skipping WorkerConfig for language:{workerDescriptionLanguage}");
+                _logger.LogInformation("{RpcWorkerConstants.FunctionWorkerRuntimeSettingName} set to {workerRuntime}. Skipping WorkerConfig for language: {}", RpcWorkerConstants.FunctionWorkerRuntimeSettingName, _workerRuntime, workerDescriptionLanguage);
```
Comment (jviau) [inline src/WebJobs.Script/Workers/Rpc/Configuration/RpcWorkerConfigFactory.cs]: Missing name for 3rd param:
```suggestion
                _logger.LogInformation("{RpcWorkerConstants.FunctionWorkerRuntimeSettingName} set to {workerRuntime}. Skipping WorkerConfig for language: {workerDescriptionLanguage}", RpcWorkerConstants.FunctionWorkerRuntimeSettingName, _workerRuntime, workerDescriptionLanguage);
```

File: src/WebJobs.Script/Workers/Rpc/WebHostRpcWorkerChannelManager.cs [hunk 5]
```diff
@@ -144,34 +144,36 @@ public async Task SpecializeAsync()
...
-                    || string.Equals(workerRuntime, RpcWorkerConstants.PythonLanguageWorkerName, StringComparison.OrdinalIgnoreCase))
-                {
-                    // Use if readonly and not v2 compatible on ~3 extension
-                    return _applicationHostOptions.CurrentValue.IsFileSystemReadOnly && !_environment.IsV2CompatibileOnV3Extension();
-                }
-                return true;
+            // Special case: node and PowerShell apps must be read-only to use the placeholder mode channel
+            // Also cannot use placeholder worker that is targeting ~3 but has backwards compatibility with V2 enabled
+            // TODO: Remove special casing when resolving https://github.com/Azure/azure-functions-host/issues/4534
+            if (string.Equals(workerRuntime, RpcWorkerConstants.NodeLanguageWorkerName, StringComparison.OrdinalIgnoreCase)
```
Comment (jviau) [inline src/WebJobs.Script/Workers/Rpc/WebHostRpcWorkerChannelManager.cs]: Not sure if you are interested in pattern matching, but you could accomplish this that way. Only drawback is it has to be constants and only ordinal works. Then you get into the issue of normalizing to upper case over lower case (Turkish i issue, but that _really_ isn't applicable here - so lowercase would be safe).

``` CSharp
if (workerRuntime.ToUpperInvariant() is "NODE" or "POWERSHELL" or "PYTHON")
```

**Review by jviau — DISMISSED** [review]
Always love seeing cleanup like this. Just the missing log param value (which I think will cause exceptions if not addressed).

File: src/WebJobs.Script/Workers/Rpc/Configuration/RpcWorkerConfigFactory.cs (hunk 4 above)
Comment (liliankasem) [reply src/WebJobs.Script/Workers/Rpc/Configuration/RpcWorkerConfigFactory.cs]: good catch!

File: src/WebJobs.Script/Workers/Rpc/WebHostRpcWorkerChannelManager.cs (hunk 5 above)
Comment (liliankasem) [reply src/WebJobs.Script/Workers/Rpc/WebHostRpcWorkerChannelManager.cs]: Doing it this way should still be okay I assume?

```
if (workerRuntime.ToLowerInvariant() is RpcWorkerConstants.NodeLanguageWorkerName
                or RpcWorkerConstants.PowerShellLanguageWorkerName
                or RpcWorkerConstants.PythonLanguageWorkerName)
```

**Review by satvu — APPROVED** [review]

File: src/WebJobs.Script/Workers/Rpc/WebHostRpcWorkerChannelManager.cs (hunk 5 above)
Comment (fabiocav) [reply src/WebJobs.Script/Workers/Rpc/WebHostRpcWorkerChannelManager.cs]: Won't the call to `ToUpperInvariant` or `ToLowerInvariant` here for the pattern matching approach cause a string copy/allocation and overall perform worse?

Comment (liliankasem) [reply src/WebJobs.Script/Workers/Rpc/WebHostRpcWorkerChannelManager.cs]: I believe Fabio did a benchmark on this and the existing approach is more performant - will revert this change in the next PR

**Pull request merged**